### ISSUE TITLE
 3D Functionality in Precast

### ISSUE THREAD

**Issue by MoussaKassim** [feiyoung/PRECAST#22]
Hello,

I am reaching out to inquire whether Precast supports 3D functionality. This feature would be highly beneficial for our current project.

Could you please confirm if Precast can be used with 3D data? If so, any additional information or documentation on how to implement this would be greatly appreciated.

Thank you in advance for your assistance.

Best regards,

**Comment by tkcaccia** [comment]
I have the same question. PRECAST uses as input a Seurat object but I cannot create a spatial object with three-dimension (e.g., xyz) with Seurat. Can PRECAST use three-dimensional spatial information? If yes, can you provide an example?

**Comment by feiyoung** [comment]
Yes, PRECAST is indeed capable of utilizing three-dimensional spatial information. Within PRECAST, this spatial information is employed to construct an adjacency matrix, which quantifies the similarities among various spots.

In our recently developed R package, ProFAST, we offer a function named 'AddAdj' that seamlessly integrates support for three-dimensional coordinates. Below, we present an illustrative example of how to incorporate this function into PRECAST, drawing upon the tutorial found at https://feiyoung.github.io/PRECAST/articles/PRECAST.BreastCancer.html. This example demonstrates the practical application of 'AddAdj' within the PRECAST framework.

Here are the codes:
install.packages(''ProFAST")
library(ProFAST)
## Create PRECASTObject.
set.seed(2022)
PRECASTObj <- CreatePRECASTObject(bc2, project = "BC2", gene.number = 2000, selectGenesMethod = "SPARK-X", premin.spots = 20, premin.features = 20, postmin.spots = 1, postmin.features = 10)
## Assume the three-dimensional cooridnates are saved in PRECASTObj@seulist[[1]]@meta.data[,c('x', 'y', 'z')]
posList <- lapply(PRECASTObj@seulist, function(x) as.matrix(@meta.data[,c('x', 'y', 'z')]))
## calculate the adjacency matrix based the three-dimensional cooridnates
PRECASTObj@AdjList <- pbapply::pblapply(posList, AddAdj, neighbors=6)
## Add a model setting in advance for a PRECASTObj object. verbose =TRUE helps outputing the
## information in the algorithm.
PRECASTObj <- AddParSetting(PRECASTObj, Sigma_equal = FALSE, verbose = TRUE, maxIter = 30)
### Given K, run PRECAST!
PRECASTObj <- PRECAST(PRECASTObj, K = 14)

**Comment by tkcaccia** [comment]
Thank you very much

**Comment by tkcaccia** [comment]
Hi,

I have been trying to use the 3D functionality of PRECAST in a MERFISH dataset with 8 tissue segments.
Here I share the link to the input data: [link](https://icgeborg-my.sharepoint.com/:u:/g/personal/cacciato_icgeb_org/EcQIhX1NPClPhDLB4SIUIogBZyR_wLKwHKXQ7hPZnOj3uw?e=HDqV4r)

Unfortunately, I did not get satisfactory results. I am not sure if I did any mistake in the script:


```

names(tissue_segments) <-  rownames(RNA)
colnames(xyz) <- c("col", "row", "z")
rownames(xyz) <- rownames(RNA)

# Create a Seurat object
library(Seurat)
seu <- CreateSeuratObject(counts = t(RNA),meta.data = data.frame(xyz))

# Set a seed for reproducibility
set.seed(543210)



# Convert the Seurat object into a list
seuList <- list(seu)


# Create a PRECAST object
PRECASTObj <- CreatePRECASTObject(seuList, project = "MERFISH", gene.number = 100,
                                  selectGenesMethod = "SPARK-X",
                                  premin.spots = 20, premin.features = 20,
                                  postmin.spots = 1, postmin.features = 10)


identifiers <- rownames(PRECASTObj@seulist[[1]]@meta.data)
colnames(xyz) <- c("x", "y", "z")
posList <- list(xyz[identifiers,])

# Extract 3D coordinates and compute the adjacency matrix

PRECASTObj@AdjList <- pbapply::pblapply(posList, AddAdj, neighbors = 15)

# Set model parameters
PRECASTObj <- AddParSetting(PRECASTObj, Sigma_equal = FALSE, verbose = TRUE, maxIter = 30)

# Run the PRECAST clustering algorithm
PRECASTObj <- PRECAST(PRECASTObj, K = 8)

# Select the optimal model
PRECASTObj <- SelectModel(PRECASTObj)


```

When I plot the results I obtained a few vertical stripes:
`plot(xyz[xyz[,3]=="-0.24",-3],col=as.numeric(PRECASTObj@resList$cluster[[1]])[xyz[,3]=="-0.24"])`

![Rplot](https://github.com/user-attachments/assets/18509c88-17db-4a2f-83dc-b88d7177acf0)
